src/cascadia/TerminalControl/ControlInteractivity.cpp

```diff
@@ -257,15 +258,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (WI_IsFlagSet(buttonState, MouseButtonState::IsRightButtonDown))
         {
-            // CopyOnSelect right click always pastes
-            if (_core->CopyOnSelect() || !_core->HasSelection())
+            if (_core->CopyOnSelect())
             {
+                // CopyOnSelect:
+                // 1. keyboard selection? --> copy the new content first
+                // 2. right click always pastes!
+                if (_core->SelectionMode() > SelectionInteractionMode::Keyboard)
+                {
+                    CopySelectionToClipboard(shiftEnabled, nullptr);
+                }
                 RequestPasteTextFromClipboard();
             }
-            else
+            else if (_core->HasSelection())
             {
+                // copy selected text
                 CopySelectionToClipboard(shiftEnabled, nullptr);
             }
+            else
+            {
+                // no selection --> paste
+                RequestPasteTextFromClipboard();
+            }

```

Comment:
Hmm.

This conditional seems like it could be simplified.

```
if (A)
  if (C)
    X
  Y
else if (B)
  X
else
  Y
```


perhaps it can't, but there's a lot of redundancy in the body...